domain separation (#139)
cfrg/draft-irtf-cfrg-hash-to-curve: draft-irtf-cfrg-hash-to-curve.md

to elliptic curve points.
This document does not cover serialization or deserialization.

### Domain separation {#term-domain-separation}

> In most cases, cryptographic protocols that use random oracles are analyzed

Reviewer: ```suggestion
Cryptographic protocols that use random oracles are often analyzed
```

> under the assumption that the random oracle answers only queries generated

Reviewer: ```suggestion
under the assumption that random oracles answer only queries generated
```

by that protocol.
In practice, this assumption may not hold: commonly, two or more protocols
may model the same hash function as a random oracle, which violates the above

> assumption if both protocols compute the hash of the same value.

Reviewer: Perhaps add:

```
That is, let R() be a random oracle used by protocols P1 and P2. If P1 and P2 ever query R with the same value x, the assumption above is violated.
```

And maybe then show how P1 and P2 would use R1 and R2 (as defined below) to address this?

Author: Yeah, I think the first paragraph is unclear in a couple ways. I'll try to address and incorporate the above.

Author: Done, I think.

A common approach to addressing this issue is called domain separation,
which allows a single random oracle to simulate multiple, independent oracles.
This is effected by ensuring that each simulated oracle sees queries that are
distinct from those seen by all other simulated oracles.
For example, to simulate two oracles R1 and R2 given a single oracle R,
one might define

    R1(x) := R("R1" || x)
    R2(x) := R("R2" || x)

In this example, "R1" and "R2" are called domain separation tags.

> Because of these domain separation tags, R1 and R2 cannot query R on

Reviewer: ```suggestion
Because of these domain separation tags, queries to R1 and R2 cannot yield identical queries to R.
```

Output: P, a point in G.

Steps:

> 1. u = hash_to_base(alpha, 2)

Reviewer: 👍

suites that specify a full set of parameters matching elliptic curves and
algorithms.

## Domain separation requirements {#domain-separation}

> When invoking hash\_to\_curve, implementors MUST use domain separation

Reviewer: ```suggestion
When invoking hash\_to\_curve in a higher-level protocol, implementors MUST use domain separation
```

({{term-domain-separation}}) to avoid interfering with other protocols
that also use the hash\_to\_curve functionality.
In addition, any protocol that uses two or more hash\_to\_curve functions

> targeting different elliptic curves MUST enforce domain separation between

Reviewer: Is the expectation that if one had two hash-to-curve functions H1 and H2 targeting the *same* curve, then H1 = H2?

Author: Yes, unless they're explicitly separated. I suppose I should clarify this point.

Author: I pushed an edit to address this.

> targeting different elliptic curves MUST enforce domain separation between

Reviewer: It'd be good to clarify why this requirement is needed. (That is, to avoid overlapping hash2base output IIUC.)

Author: I pushed an edit to address this.

the two functions if those functions are modeled in the protocol as
independent random oracles.
Finally, protocols that use encode\_to\_curve SHOULD use domain separation

> if possible, but it is not required in this case.

Reviewer: ```suggestion
if possible, though it is not required in this case.
```

Care is required when choosing a domain separation tag.
Implementors SHOULD observe the following guidelines:

1. Tags should be prepended to the value being hashed, as in the example
   in {{term-domain-separation}}.

2. Tags should have fixed length, or should be encoded in a way that makes
   the length of a given tag unambiguous.
   If a variable-length tag is used, it should be prefixed with a
   fixed-length field that encodes the length of the tag.

3. Tags should begin with a fixed protocol identification string.
   Ideally, this identification string should be unique to the protocol.

4. Tags should include a protocol version number.

5. For protocols that support multiple ciphersuites, tags should include
   a ciphersuite identifier.

As an example, consider a fictional key exchange protocol named Quux.
A reasonable choice of tag might be "QUUX-V\<xx\>-CS\<yy\>", where \<xx\> and \<yy\>

> are two-digit numbers indicating the version and ciphersuite, respectively.

Reviewer: Would it help to give an example of a protocol that needs domain separation internally, too?

Author: I pushed an edit to address this.

Inputs:
- msg is the message to hash.
- ctr is either 0, 1, or 2.
  This is used to efficiently create independent
  instances of hash_to_base (see discussion above).

Output: u, an element in F.

Steps:

> 1. m' = "HASH-TO-CURVE" || H(msg) || I2OSP(ctr, 1)

Reviewer 2: Would we consider making HASH-TO-CURVE shorter (e..g "H2C"/ "HTC"), and potentially truncating H(msg) such that we can fit m' into a single block in the loop below?

E.g. the input to the hash in the loop becomes `"H2C" || H(msg)[..len(H) - 6] || I2OSP(ctr, 1)  || I2OSP(i, 1) || I2OSP(j, 1)`?

Reviewer: Perhaps in a separate issue?

Author: we'd have to truncate a reasonable amount to avoid going past one block, because SHA-256 adds at least 65 bits of padding to the end of the message. Specifically, the padding is

- append a single '1' bit
- append k '0' bits where 0 <= k < 512 and bitlen(msg) + 1 + k + 64 is divisible by 512
- append a 64-bit representation of bitlen(msg)

so it would have to be something like H(msg)[..len(H) - 15], which is pretty extreme.

Also, we might want to resolve #137 first, since that would moot this discussion.
